mlir/lib/Pass/Pass.cpp

```diff
@@ -513,6 +513,21 @@ LogicalResult OpToOpPassAdaptor::run(Pass *pass, Operation *op,
   };
   pass->passState.emplace(op, am, dynamicPipelineCallback);
 
+  // Get any op dependent dialects.
+  DialectRegistry dependentDialects;
+  pass->getOpDependentDialects(op, dependentDialects);
+  MLIRContext &context = pass->getContext();
+  // Only append a non-empty non-trivial registry.
+  if (!dependentDialects.empty() &&
+      !dependentDialects.isSubsetOf(context.getDialectRegistry())) {
+    context.executeCriticalSection([&]() {
+      // Load the dialects.
+      context.appendDialectRegistry(dependentDialects);
+      for (StringRef name : dependentDialects.getDialectNames())
+        context.getOrLoadDialect(name);
```

Comment:
This seems broken to me: it'll load dialect (and as such change the behavior of the system) while running a pass manager. This will lead to a potentially non-deterministic compiler.

Reply:
Imagine: `pass-pipeline=func.func(PassA, PassB)`
Now with a module containing `func.func @func1` and  `func.func @func2`.

Assume that PassA would `dyn_cast<FooOpInterface>` and this interface is loaded by a dialect you're trying to load in PassB here.

Some of the possible executions are:

```
PassA(@func1) // dyn_cast<FooOpInterface>
PassA(@func2) // dyn_cast<FooOpInterface>
PassB(@func1) // load FooOpInterface
PassB(@func2) // load FooOpInterface
```

```
PassA(@func1) // dyn_cast<FooOpInterface>
PassB(@func1) // load FooOpInterface
PassA(@func2) // dyn_cast<FooOpInterface>
PassB(@func2) // load FooOpInterface
```


Now you see here that in the first case, PassA never sees the `FooOpInterface` because it's not loaded, while in the second case it is loaded before PassA operates on func2.

Reply:
Isn't that in the realm of a bug or undefined behavior? If pass A required `FooOpInterface`, then it should have loaded it before running.
What I meant with my comment is, in a fully correct setup, I think loading a dialect shouldn't introduce non-determinism.